tensorflow/core/kernels/data/concatenate_dataset_op_test.cc

```diff
@@ -0,0 +1,596 @@
+    inputs.emplace_back(&tensor);
+  }
+
+  std::unique_ptr<OpKernel> dataset_kernel;
+  TF_ASSERT_OK(CreateConcatenateDatasetKernel(
+      expected_output_dtypes, expected_output_shapes, &dataset_kernel));
+  std::unique_ptr<OpKernelContext> dataset_kernel_ctx;
+  TF_ASSERT_OK(CreateConcatenateDatasetContext(dataset_kernel.get(), &inputs,
+                                               &dataset_kernel_ctx));
+  DatasetBase *dataset;
+  TF_ASSERT_OK(
+      CreateDataset(dataset_kernel.get(), dataset_kernel_ctx.get(), &dataset));
+  core::ScopedUnref scoped_unref(dataset);
+
+  EXPECT_OK(VerifyTypesMatch(dataset->output_dtypes(), expected_output_dtypes));
+}
+
+INSTANTIATE_TEST_CASE_P(ConcatenateDatasetOpTest, DatasetOutputDtypesTest,
+                        ::testing::ValuesIn(TestCases));
+
+struct DatasetOutputShapesTest : ConcatenateDatasetOpTest,
+                                 ::testing::WithParamInterface<TestParam> {};
+
+TEST_P(DatasetOutputShapesTest, DatasetOutputShapes) {
+  int thread_num = 2, cpu_num = 2;
+  TF_ASSERT_OK(InitThreadPool(thread_num));
+  TF_ASSERT_OK(InitFunctionLibraryRuntime({}, cpu_num));
+
+  std::vector<std::vector<Tensor>> tensor_vectors = GetParam().input_tensors;
+  std::vector<Tensor> expected_outputs = GetParam().expected_outputs;
+  std::vector<PartialTensorShape> expected_output_shapes =
+      GetParam().expected_output_shapes;
+  DataTypeVector expected_output_dtypes = GetParam().expected_output_dtypes;
+  std::vector<Tensor> tensor_slice_dataset_tensors;
+  TF_ASSERT_OK(CreateTensorSliceDatasetTensors(tensor_vectors,
+                                               &tensor_slice_dataset_tensors));
+  gtl::InlinedVector<TensorValue, 4> inputs;
+  for (auto &tensor : tensor_slice_dataset_tensors) {
+    inputs.emplace_back(&tensor);
+  }
+
+  std::unique_ptr<OpKernel> dataset_kernel;
+  TF_ASSERT_OK(CreateConcatenateDatasetKernel(
+      expected_output_dtypes, expected_output_shapes, &dataset_kernel));
+  std::unique_ptr<OpKernelContext> dataset_kernel_ctx;
+  TF_ASSERT_OK(CreateConcatenateDatasetContext(dataset_kernel.get(), &inputs,
+                                               &dataset_kernel_ctx));
+  DatasetBase *dataset;
+  TF_ASSERT_OK(
+      CreateDataset(dataset_kernel.get(), dataset_kernel_ctx.get(), &dataset));
+  core::ScopedUnref scoped_unref(dataset);
+
+  EXPECT_OK(
+      VerifyShapesCompatible(dataset->output_shapes(), expected_output_shapes));
+}
+
+INSTANTIATE_TEST_CASE_P(ConcatenateDatasetOpTest, DatasetOutputShapesTest,
+                        ::testing::ValuesIn(TestCases));
+
+struct DatasetCardinalityTest : ConcatenateDatasetOpTest,
+                                ::testing::WithParamInterface<TestParam> {};
+
+TEST_P(DatasetCardinalityTest, Cardinality) {
+  int thread_num = 2, cpu_num = 2;
+  TF_ASSERT_OK(InitThreadPool(thread_num));
+  TF_ASSERT_OK(InitFunctionLibraryRuntime({}, cpu_num));
+
+  std::vector<std::vector<Tensor>> tensor_vectors = GetParam().input_tensors;
```

Comment:
const-reference, likewise for other non primitive params in the rest of the tests.